src/protovm/rw_proto.h

```diff
@@ -0,0 +1,158 @@
+//               │            │
+//               │            │
+//               │   *********▼********
+//               │   *                *
+//               │   * Mapped         *
+//               │   * RepeatedField  *
+//               │   *                *
+//               │   ******************  Map<key, OwnedPtr<Node>>
+//               │           │ │
+//               │       ┌───┘ └────┐
+//               │       │          │
+//               │  ┌────▼────┐ ┌───▼───┐
+//               │  │         │ │       │
+//               │  │ Message │ │ Bytes │
+//               │  │         │ │       │
+//               │  └────┬────┘ └───────┘
+//               │       ▼
+//               │      ...
+//               │
+//
+//       ********▼*********
+//       *                *
+//       * Indexed        *
+//       * RepeatedField  *
+//       *                *
+//       ******************  Map<index, OwnedPtr<Node>>
+//             ││││
+//      ┌──────┘││└────────────────────┐
+//      │       │└───────────┐         │
+//      │       └──┐         │         │
+//      │          │         │         │
+// ┌────▼────┐ ┌───▼───┐ ┌───▼───┐ ┌───▼────┐
+// │         │ │       │ │       │ │        │
+// │ Message │ │ Empty │ │ Bytes │ │ Scalar │
+// │         │ │       │ │       │ │        │
+// └────┬────┘ └───────┘ └───────┘ └────────┘
+//      ▼
+//     ...
+//
+//
+// ┌─┐
+// │ │  Data node: corresponds to an actual field/message in the proto schema
+// └─┘             and eventually becomes part of the final serialized proto.
+//
+// ***
+// * *  Structural node: used to organize and provide access to message fields.
+// ***                   It is an internal helper and doesn't appear in the
+//                       final serialized proto.
+//
+//
+// Message: Represents a protobuf message. It maps field_id to corresponding
+//          node/value. It allows accessing fields as well as adding and
+//          removing them.
+//
+//
+// Empty: Represents an empty or uninitialized node. It is used as a
+//        placeholder before a node's type is determined. E.g. between an
+//        "enter field" and a "set field value" operation.
+//
+//
+// Bytes: Stores a sequence of bytes, typically representing a length-delimited
+//        protobuf field.
+//
+//
+// Scalar: Stores a scalar value.
+//
+//
+// IndexedRepeatedField: Represents a protobuf repeated field where elements
+//                       are accessed by index.
+//
+//
+// MappedRepeatedField: Represents a protobuf repeated field where elements
+//                      are accessed by key.
+class RwProto {
+ public:
+  using Cursor = RwProtoCursor;
+
+  explicit RwProto(Allocator* allocator);
+  ~RwProto();
+  Cursor GetRoot();
+  std::string SerializeAsString() const;
+
+ private:
+  void SerializeField(uint32_t field_id,
+                      const Node& node,
+                      protozero::Message* proto) const;
+
+  Allocator* allocator_;
```

Comment:
either:
1)make it Allocator* const allocator_;
2) make it = nullptr

to enforce that nobody will forget to initialize this